server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -234,14 +266,28 @@ class AsyncAction {
             this.request = request;
             this.listener = listener;
             if (request.concreteNodes() == null) {
-                resolveRequest(request, clusterService.state());
-                assert request.concreteNodes() != null;
+                if (optimizedTransportActionNames.contains(request.getClass().getSimpleName())) {
```

Comment:
Can you add some tests for this? Also, have you executed all existing tests and made sure that they are passing after your change. If yes, then please confirm this in the PR description.

Reply:
Please make sure that node left and node added scenarios are covered properly.

Reply:
Yes, I have modified the draft PR approach. Will make sure I add all the necessary tests covering different scenarios.